spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Dropout.scala

```diff
@@ -49,45 +53,48 @@ class Dropout[T: ClassTag](
       results = new Array[Future[Unit]](Engine.model.getPoolSize)
     }
     if (train) {
-      noise.resizeAs(input)
-      if (input.isContiguous()) {
-        val noiseData = noise.storage().array()
-        var taskSize = noise.nElement() / Engine.model.getPoolSize
-        var extraTask = noise.nElement() % Engine.model.getPoolSize
-        var allocated = 0
-        val offset = this.output.storageOffset() - 1
-        val data = this.output.storage.array()
-        var i = 0
-        while (allocated < noise.nElement()) {
-          val start = allocated
-          allocated += taskSize
-          if (extraTask > 0) {
-            allocated += 1
-            extraTask -= 1
-          }
-          val end = allocated
-          results(i) = Engine.model.invoke(() => {
-            var k = start
-            while (k < end) {
-              noiseData(k) = if (RNG.bernoulli(1 - p)) {
-                if (scale) {
-                  data(offset + k) = ev.divide(data(offset + k), ev.fromType[Double](1 - p))
-                  ev.fromType[Double](1.0 / (1 - p))
+      if (!isLazy || flag(0)) {
+        noise.resizeAs(input)
+        if (input.isContiguous()) {
+          val noiseData = noise.storage().array()
+          var taskSize = noise.nElement() / Engine.coreNumber()
+          var extraTask = noise.nElement() % Engine.coreNumber()
+          var allocated = 0
+          val offset = this.output.storageOffset() - 1
+          val data = this.output.storage.array()
+          var i = 0
+          while (allocated < noise.nElement()) {
+            val start = allocated
+            allocated += taskSize
+            if (extraTask > 0) {
+              allocated += 1
+              extraTask -= 1
+            }
+            val end = allocated
+            results(i) = Engine.model.invoke(() => {
+              var k = start
+              while (k < end) {
+                noiseData(k) = if (RNG.bernoulli(1 - p)) {
+                  if (scale) {
+                    data(offset + k) = ev.divide(data(offset + k), ev.fromType[Double](1 - p))
+                    ev.fromType[Double](1.0 / (1 - p))
+                  } else {
+                    ev.fromType[Int](1)
+                  }
                 } else {
-                  ev.fromType[Int](1)
+                  data(offset + k) = ev.fromType[Int](0)
+                  ev.fromType[Int](0)
                 }
-              } else {
-                data(offset + k) = ev.fromType[Int](0)
-                ev.fromType[Int](0)
+
+                k += 1
               }
+            })
+            i += 1
+          }
 
-              k += 1
-            }
-          })
-          i += 1
+          Engine.model.sync(results)
+          flag(0) = false
         }
-
-        Engine.model.sync(results)
         this.output
       } else {
```

Comment:
Yes, that was the old version not passed the dropout test with a logic problem. I uploaded the new version code just now and fixed the problem.